src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java

```diff
@@ -312,6 +391,205 @@ public void processElement(final ProcessContext context) throws Exception {
+                            final GATKRead record = context.element();
+                            final SAMFileHeader h = context.sideInput(header);
+                            final String library = LibraryIdGenerator.getLibraryName(h, record.getReadGroup());
+                            DuplicationMetrics metrics = new DuplicationMetrics();
+                            metrics.LIBRARY = library;
+                            // These need to be initialized so that the DuplicationMetrics can be serialized safely.
+                            metrics.PERCENT_DUPLICATION = 0.0;
+                            metrics.ESTIMATED_LIBRARY_SIZE = 0L;
+                            // We have already filtered out secondary and supplementary reads in the partition transform.
+                            if (record.isUnmapped()) {
+                                ++metrics.UNMAPPED_READS;
+                            } else if (!record.isPaired() || record.mateIsUnmapped()) {
+                                ++metrics.UNPAIRED_READS_EXAMINED;
+                            } else {
+                                ++metrics.READ_PAIRS_EXAMINED;
+                            }
+
+                            if (record.isDuplicate()) {
+                                if (!record.isPaired() || record.mateIsUnmapped()) {
+                                    ++metrics.UNPAIRED_READ_DUPLICATES;
+                                } else {
+                                    ++metrics.READ_PAIR_DUPLICATES;
+                                }
+                            }
+                            if (record.hasAttribute(OPTICAL_DUPLICATE_ATTRIBUTE_NAME) &&
+                                record.getAttributeAsInteger(OPTICAL_DUPLICATE_ATTRIBUTE_NAME) == 1) {
+                                ++metrics.READ_PAIR_OPTICAL_DUPLICATES;
+                            }
+                            final KV<String, DuplicationMetrics> kv = KV.of(library, metrics);
+                            context.output(kv);
+                        }
+                    });
+        }
+
+
+        public static class CombineMetricsFn implements SerializableFunction<Iterable<DuplicationMetrics>, DuplicationMetrics> {
+            private static final long serialVersionUID = 1l;
+
+            @Override
+            public DuplicationMetrics apply(Iterable<DuplicationMetrics> input) {
+                DuplicationMetrics metricsSum = new DuplicationMetrics();
+                int num = 0;
+                for (final DuplicationMetrics metrics : input) {
+                    if (metricsSum.LIBRARY == null) {
+                        metricsSum.LIBRARY = metrics.LIBRARY;
+                    }
+                    // This should never happen, as we grouped by key using library as the key.
+                    if (!metricsSum.LIBRARY.equals(metrics.LIBRARY)) {
+                        return null;
```

Comment:
The issue is that apply doesn't throw an error, so I wasn't sure what else to do.